src/main/java/uk/gov/cshr/config/CustomAuthenticationSuccessHandler.java

```diff
@@ -0,0 +1,117 @@
+import uk.gov.cshr.service.EmailUpdateService;
+import uk.gov.cshr.service.security.IdentityDetails;
+import uk.gov.cshr.service.security.IdentityService;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpSession;
+import java.io.IOException;
+import java.util.Optional;
+
+@Slf4j
+public class CustomAuthenticationSuccessHandler
+        implements AuthenticationSuccessHandler {
+
+    @Value("${emailUpdate.invalidDomainUrl}")
+    private String invalidDomainUrl;
+
+    @Value("${lpg.uiUrl}")
+    private String lpgUiUrl;
+
+    @Autowired
+    private RedirectStrategy redirectStrategy;
+
+    @Autowired
+    private AgencyTokenService agencyTokenService;
+
+    @Autowired
+    private IdentityService identityService;
+
+    @Autowired
+    private EmailUpdateService emailUpdateService;
+
+    @Autowired
+    private IdentityRepository identityRepository;
+
+    @Override
+    public void onAuthenticationSuccess(HttpServletRequest request,
+                                        HttpServletResponse response, Authentication authentication)
+            throws IOException {
+        handle(request, response, authentication);
+        clearAuthenticationAttributes(request);
+    }
+
+    protected void handle(HttpServletRequest request,
+                          HttpServletResponse response, Authentication authentication)
+            throws IOException {
+
+        String targetUrl = determineTargetUrl(authentication);
+
+        if (response.isCommitted()) {
+            log.debug(
+                    "Response has already been committed. Unable to redirect to "
+                            + targetUrl);
+            return;
+        }
+
+        log.debug("Redirecting to " + targetUrl);
+        redirectStrategy.sendRedirect(request, response, targetUrl);
+    }
+
+    protected String determineTargetUrl(Authentication authentication) {
+        String targetURL = lpgUiUrl;
+        // GET FLAG FROM IDENTITY REPO (NOT SPRING AUTH)
+        IdentityDetails identityDetails = (IdentityDetails) authentication.getPrincipal();
+        Identity identity = identityDetails.getIdentity();
+
+        Optional<Identity> testIdentityOptional = identityRepository.findFirstByUid(identity.getUid());
+        if(testIdentityOptional.isPresent()) {
+            Identity testIdentity = testIdentityOptional.get();
+            log.debug("test identity=" + testIdentity);
+        }
+
+        boolean hasUserRecentlyUpdatedTheirEmail = identityService.getRecentlyUpdatedEmailFlag(identity);
+        if(hasUserRecentlyUpdatedTheirEmail) {
+            log.debug("users email has recently been updated");
+            targetURL = workoutIfUserShouldUseAnAgencyToken(identity);
+        }
+
+        return targetURL;
+    }
+
+    private String workoutIfUserShouldUseAnAgencyToken(Identity identity) {
+        String domain = identityService.getDomainFromEmailAddress(identity.getEmail());
```

Comment:
This should probably be renamed to 'getTargetUrl()`

Reply:
Updated